### PR TITLE
Feature file storage

### PR DESCRIPTION
**Pull request by minsikl**
Three files are added to provide file storage resource as follows:
- resource_softlayer_file_storage.go
- resource_softlayer_file_storage_test.go
- softlayer_file_storage.md

File storage resource provides functions to create endurance file storage and performance file storage.
 - Users can define endurance or performance storage with various size and IOPS.
 - Endurance storage supports snapshot configuration.
 - Access control can be configured with `allowed_virtual_guest_ids`, `allowed_hardware_ids`, `allowed_subnets`, and `allowed_ip_addresses`.

### CODE REVIEW

File: softlayer/resource_softlayer_file_storage.go [hunk 1]
```diff
@@ -0,0 +1,941 @@
...
+				"storageProtocolCategoryCode": "performance_storage_iscsi",
+			},
+			EnduranceType: {
+				"storagePackageType":          StorageEndurancePackageType,
+				"iopsCategoryCode":            "storage_tier_level",
+				"storageProtocolCategoryCode": "storage_block",
+			},
+		},
+	}
+)
```
Comment (renier) [inline softlayer/resource_softlayer_file_storage.go:81]: @minsikl Should any of these mappings exist in softlayer-go helpers?

File: softlayer/resource_softlayer_file_storage.go [hunk 2]
```diff
@@ -0,0 +1,941 @@
...
+		PackageId: pkg.Id,
+		Location:  sl.String(strconv.Itoa(*dc.Id)),
+		Prices:    targetItemPrices,
+		Quantity:  sl.Int(1),
+	}
+
+	return productOrderContainer, nil
+}
+
+func findStorageByOrderId(sess *session.Session, orderId int) (datatypes.Network_Storage, error) {
```
Comment (renier) [inline softlayer/resource_softlayer_file_storage.go:519]: Candidate for softlayer-go helper?

**Review by renier — COMMENTED** [review]
Looks good. My only overall comment is whether any code here should move to softlayer-go helper if it could be helpful to others.

File: softlayer/resource_softlayer_file_storage.go [hunk 3]
```diff
@@ -0,0 +1,941 @@
...
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/session"
+	"github.com/softlayer/softlayer-go/sl"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	StoragePerformancePackageType = "ADDITIONAL_SERVICES_PERFORMANCE_STORAGE"
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: Why export these constants? Could they also be used for ordering block storage? Perhaps we should consider creating a package just for storage?

File: softlayer/resource_softlayer_file_storage.go [hunk 4]
```diff
@@ -0,0 +1,941 @@
...
+		",serviceResourceName,allowedIpAddresses,allowedSubnets,allowedVirtualGuests,snapshotCapacityGb,osType"
+	itemMask        = "id,capacity,description,units,keyName,prices[id,categories[id,name,categoryCode],capacityRestrictionMinimum,capacityRestrictionMaximum,locationGroupId]"
+	EnduranceType   = "Endurance"
+	PerformanceType = "Performance"
+	FileStorage     = "FILE_STORAGE"
+	BlockStorage    = "BLOCK_STORAGE"
+)
+
+var (
+	enduranceIopsMap = map[float64]string{
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: Could you leave a comment here explaining how these values fit into the SoftLayer ordering scheme? It would be very helpful for future developers.

File: softlayer/resource_softlayer_file_storage.go [hunk 5]
```diff
@@ -0,0 +1,941 @@
...
+				"iopsCategoryCode":            "performance_storage_iops",
+				"storageProtocolCategoryCode": "performance_storage_nfs",
+			},
+			EnduranceType: {
+				"storagePackageType":          StorageEndurancePackageType,
+				"iopsCategoryCode":            "storage_tier_level",
+				"storageProtocolCategoryCode": "storage_file",
+			},
+		},
+		BlockStorage: {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: This resource is only for ordering FileStorage, but it has mappings for BlockStorage as well?

Perhaps a `/lib/network_storage` package would be useful for storing common network storage utility functions and constants.

File: softlayer/resource_softlayer_file_storage.go [hunk 6]
```diff
@@ -0,0 +1,941 @@
...
+
+	// Wait for storage availability
+	_, err = WaitForStorageAvailable(d, meta)
+
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for storage (%s) to become ready: %s", d.Id(), err)
+	}
+
+	// SoftLayer changes the device ID after completion of provisioning. It is necessary to refresh device ID.
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:226]: Good to know.

File: softlayer/resource_softlayer_file_storage.go [hunk 7]
```diff
@@ -0,0 +1,941 @@
...
+	d.Set("capacity", *storage.CapacityGb)
+	d.Set("volumename", *storage.Username)
+	d.Set("hostname", *storage.ServiceResourceBackendIpAddress)
+	d.Set("iops", iops)
+	if storage.SnapshotCapacityGb != nil {
+		snapshotCapacity, _ := strconv.Atoi(*storage.SnapshotCapacityGb)
+		d.Set("snapshot_capacity", snapshotCapacity)
+	}
+
+	// Parse data center short name from ServiceResourceName
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: Can you list an example? This might also be a good area to write a unit test for this regex. @renier is there a template for writing unit tests?

Comment (Freyert) [reply softlayer/resource_softlayer_file_storage.go]: I would also say a lot of the code below this comment is also easily unit testable. Breaking out the functionality that creates these lists and testing it individually will be incredibly useful in the future.

It may seem like a very pain staking thing to do right now, but if any edge cases or errors crop up around these things it will be nice to have a starting point for unit tests.

File: softlayer/resource_softlayer_file_storage.go [hunk 8]
```diff
@@ -0,0 +1,941 @@
...
+	switch storageType {
+	case EnduranceType:
+		return enduranceIopsMap[iops], nil
+	case PerformanceType:
+		return fmt.Sprintf("%.f_IOPS", iops), nil
+	}
+	return "", fmt.Errorf("Invalid storageType %s.", storageType)
+}
+
+func getPrice(productItems []datatypes.Product_Item, keyName string, categoryCode string, capacityRestrictionType string, capacityRestriction int) (datatypes.Product_Item_Price, error) {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:633]: This function is begging for a unit test. Unit tests here would go a long way towards explaining what this function is capable of, and it would allow developers to extend it without fear of breaking it.

File: softlayer/resource_softlayer_file_storage.go [hunk 9]
```diff
@@ -0,0 +1,941 @@
...
+					}
+				}
+			}
+		}
+	}
+	return datatypes.Product_Item_Price{},
+		fmt.Errorf("No product items matching with keyName %s and categoryCode %s could be found", keyName, categoryCode)
+}
+
+func getIops(storage datatypes.Network_Storage, storageType string) (float64, error) {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:677]: This function is also begging for some unit tests.

File: softlayer/resource_softlayer_file_storage.go [hunk 10]
```diff
@@ -0,0 +1,941 @@
...
+	}
+	return 0, fmt.Errorf("Invalied storage type %s", storageType)
+}
+
+func updateAllowedIpAddresses(d *schema.ResourceData, sess *session.Session, storage datatypes.Network_Storage) error {
+	id := *storage.Id
+	newIps := d.Get("allowed_ip_addresses").(*schema.Set).List()
+
+	// Add new allowed_ip_addresses
+	for _, newIp := range newIps {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:707]: This logic should be extracted to another function and unit tested. The rest of this function is calls out to a server and is better covered by the acceptance tests.

File: softlayer/resource_softlayer_file_storage.go [hunk 11]
```diff
@@ -0,0 +1,941 @@
...
+					},
+				})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Remove deleted allowed_hardware_ids
+	for _, oldAllowedIpAddresses := range storage.AllowedIpAddresses {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:741]: Also extract and unit test this logic.

File: softlayer/resource_softlayer_file_storage.go [hunk 12]
```diff
@@ -0,0 +1,941 @@
...
+	}
+	return nil
+}
+
+func updateAllowedSubnets(d *schema.ResourceData, sess *session.Session, storage datatypes.Network_Storage) error {
+	id := *storage.Id
+	newSubnets := d.Get("allowed_subnets").(*schema.Set).List()
+
+	// Add new allowed_subnets
+	for _, newSubnet := range newSubnets {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:771]: Good logic to unit test here.

File: softlayer/resource_softlayer_file_storage.go [hunk 13]
```diff
@@ -0,0 +1,941 @@
...
+					},
+				})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Remove deleted allowed_subnets
+	for _, oldAllowedSubnets := range storage.AllowedSubnets {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:810]: Candidate for a unit test.

File: softlayer/resource_softlayer_file_storage.go [hunk 14]
```diff
@@ -0,0 +1,941 @@
...
+	}
+	return nil
+}
+
+func updateAllowedVirtualGuestIds(d *schema.ResourceData, sess *session.Session, storage datatypes.Network_Storage) error {
+	id := *storage.Id
+	newIds := d.Get("allowed_virtual_guest_ids").(*schema.Set).List()
+
+	// Add new allowed_virtual_guest_ids
+	for _, newId := range newIds {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:847]: Candidate for a unit test.

File: softlayer/resource_softlayer_file_storage.go [hunk 15]
```diff
@@ -0,0 +1,941 @@
...
+	}
+	return nil
+}
+
+func updateAllowedHardwareIds(d *schema.ResourceData, sess *session.Session, storage datatypes.Network_Storage) error {
+	id := *storage.Id
+	newIds := d.Get("allowed_hardware_ids").(*schema.Set).List()
+
+	// Add new allowed_hardware_ids
+	for _, newId := range newIds {
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go:901]: There seems to be a lot of comment logic between all these `updateAllowed*`.

File: softlayer/resource_softlayer_file_storage.go [hunk 16]
```diff
@@ -0,0 +1,941 @@
...
+				provisionedIops, err := strconv.Atoi(*property.Value)
+				if err != nil {
+					return 0, err
+				}
+				return float64(provisionedIops) / float64(*storage.CapacityGb), nil
+			}
+		}
+	case PerformanceType:
+		if storage.Iops == nil {
+			return 0, fmt.Errorf("Failed to retrive iops information.")
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: spelling error: retrive -> retrieve.

File: softlayer/resource_softlayer_file_storage.go [hunk 17]
```diff
@@ -0,0 +1,941 @@
...
+		}
+		iops, err := strconv.Atoi(*storage.Iops)
+		if err != nil {
+			return 0, err
+		}
+		return float64(iops), nil
+	}
+	return 0, fmt.Errorf("Invalied storage type %s", storageType)
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: Spelling error: Invalied -> Invalid.

File: softlayer/resource_softlayer_file_storage.go [hunk 18]
```diff
@@ -0,0 +1,941 @@
...
+			}, sl.Bool(false))
+	case PerformanceType:
+		receipt, err = services.GetProductOrderService(sess).PlaceOrder(
+			&datatypes.Container_Product_Order_Network_PerformanceStorage_Nfs{
+				Container_Product_Order_Network_PerformanceStorage: datatypes.Container_Product_Order_Network_PerformanceStorage{
+					Container_Product_Order: storageOrderContainer,
+				},
+			}, sl.Bool(false))
+	default:
+		return fmt.Errorf("Error during creation of storage: Invalied storageType %s", storageType)
```
Comment (Freyert) [inline softlayer/resource_softlayer_file_storage.go]: Spelling Error: Invalied -> Invalid.

**Review by Freyert — COMMENTED** [review]
I'll second @renier about there being a lot of great functionality here that we could reuse.

My thought is that it might be easier to maintain these functions in a separate package with in this package until the functions mature and are found to be correct.

There were a lot of functions, and parts of functions here that were strictly synchronous logic (no I/O). "Pure functions" like these are great for unit tests though they seem like trivial logic. I think it would really help us all out if we started building some unit test buckets and stacking up some code coverage if we can. I am definitely willing to help refactor and test.

This is some great work. I pointed out a few spelling errors here and there though :). I would say they spelling errors would be great to have fixed, and perhaps we can follow up with a PR for unit tests and refactoring.

File: softlayer/resource_softlayer_file_storage.go (hunk 2 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go:519]: Let me think about creating a more generic module.

File: softlayer/resource_softlayer_file_storage.go (hunk 3 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: They can be used for ordering both block storage and file storage.

File: softlayer/resource_softlayer_file_storage.go (hunk 5 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: `buildStorageProductOrderContainer` function supports both block_storage and file_storage.

File: softlayer/resource_softlayer_file_storage.go (hunk 3 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: Changed constants to private values.

File: softlayer/resource_softlayer_file_storage.go (hunk 4 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: Added comments.

File: softlayer/resource_softlayer_file_storage.go (hunk 18 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: Thank you. Fixed.

File: softlayer/resource_softlayer_file_storage.go (hunk 6 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go:226]: Only happened in specific data centers.

File: softlayer/resource_softlayer_file_storage.go (hunk 7 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: Added an example.

File: softlayer/resource_softlayer_file_storage.go (hunk 16 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: Fixed.

File: softlayer/resource_softlayer_file_storage.go (hunk 17 above)
Comment (minsikl) [reply softlayer/resource_softlayer_file_storage.go]: Fixed.
